### PR TITLE
Add goodness of fit test helpers for testing distributions

### PR DESCRIPTION
**Pull request by fritzo**
Addresses #2658 

This adds a PyTorch port of the [goftests](https://github.com/posterior/goftests) library, enabling easy statistical testing between distributions' `.sample()` and `.log_prob()` methods. I've had good luck with these statistical tests in a number of past projects, including Loom and Gamalon's PPL.

The main function is `auto_goodness_of_fit(samples, probs)` which returns a p-value that should be Uniform(0,1)-distributed for good data but should be close to zero for bad data. This well-calibrated statistic allows precise balancing and tuning of failure rate across a suite of tests:
```py
TEST_FAILURE_RATE = 1 / 100  # For 1 in 100 chance of spurious failure.

def test_my_distribution():
    d = MyDistribution()
    samples = d.sample([10000])
    probs = d.log_prob(samples).exp()
    gof = auto_goodness_of_fit(samples, probs)
    assert gof > TEST_FAILURE_RATE
```

## Tested

The following tests take less than 4 seconds total:

- [x] added a generic test helper to `test_distributions.py`
- [x] added two tests to `test_von_mises.py` exercising strict submanifold behavior
- [x] ported the unit test for `chi2sf()`

### CODE REVIEW

File: pyro/distributions/testing/gof.py [hunk 1]
```diff
@@ -0,0 +1,280 @@
+# Copyright (c) 2014, Salesforce.com, Inc.  All rights reserved.
```
Comment (fritzo) [inline pyro/distributions/testing/gof.py:1]: This file is ported from https://github.com/posterior/goftests/blob/master/goftests/__init__.py
Differences include:
- this restricts to continuous distributions
- this ports to PyTorch and Python 3

File: pyro/distributions/testing/special.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
+# Copyright (c) 2014, Salesforce.com, Inc.  All rights reserved.
```
Comment (fritzo) [inline pyro/distributions/testing/special.py:1]: This file is ported from https://github.com/posterior/goftests/blob/master/goftests/utils.py
Differences include:
- port to Python 3

File: pyro/distributions/testing/gof.py [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+    counts,
+    *,
+    total_count=None,
+    plot=False,
+):
+    """
+    Pearson's chi^2 test, on possibly truncated data.
+    https://en.wikipedia.org/wiki/Pearson%27s_chi-squared_test
+
+    :param Tensor probs: Vector of probabilities.
```
Comment (neerajprad) [inline pyro/distributions/testing/gof.py:92]: nit: I think you may need to import `Tensor` or use `torch.Tensor` here?

**Review by neerajprad — APPROVED** [review]
Thanks for adding these, it is is a great addition to distribution testing utils! I have not checked the math here which should be fine given the lineage of the original codebase, but the tests and documentation look great.

**Comment by fritzo** [comment]
Thanks for reviewing, @neerajprad !

**Review by neerajprad — APPROVED** [review]

**Pull request merged**